cs/src/Connections/RelayTunnelConnector.cs

```diff
@@ -48,169 +48,220 @@ public RelayTunnelConnector(IRelayClient relayClient)
         SshDisconnectReason disconnectReason;
         Exception? exception;
 
+        relayClient.StartConnecting();
+        using var cts = CancellationTokenSource.CreateLinkedTokenSource(
+            this.relayClient.DisposeToken,
+            cancellation);
+
         while (true)
         {
-            cancellation.ThrowIfCancellationRequested();
             attempt++;
             isDelayNeeded = true;
             Stream? stream = null;
             errorDescription = null;
             disconnectReason = SshDisconnectReason.ConnectionLost;
             exception = null;
+            var isFinalAttempt = false;
             try
             {
-                // TODO: check tunnel access token expiration and try refresh it if it's expired.
-                stream = await this.relayClient.CreateSessionStreamAsync(cancellation);
-                await this.relayClient.ConfigureSessionAsync(stream, isReconnect, cancellation);
+                try
+                {
+                    cts.Token.ThrowIfCancellationRequested();
 
-                stream = null;
-                disconnectReason = SshDisconnectReason.None;
-                return;
-            }
-            catch (UnauthorizedAccessException uaex) // Tunnel access token validation failed.
-            {
-                await RefreshTunnelAccessTokenAsync(uaex);
-            }
-            catch (SshReconnectException srex)
-            {
-                errorDescription = srex.Message;
-                disconnectReason = SshDisconnectReason.ProtocolError;
-                exception = srex;
-                isDelayNeeded = false;
-                isReconnect = false;
-            }
-            catch (SshConnectionException scex) when (scex.DisconnectReason == SshDisconnectReason.ConnectionLost)
-            {
-                // Recoverable
-                errorDescription = scex.Message;
-            }
-            catch (SshConnectionException scex)
-            {
-                // All other SSH connection exceptions are not recoverable.
-                disconnectReason = scex.DisconnectReason != SshDisconnectReason.None ? scex.DisconnectReason : SshDisconnectReason.ByApplication;
-                throw exception = new TunnelConnectionException($"Failed to start tunnel SSH session: {scex.Message}", scex);
-            }
-            catch (WebSocketException wse)
-            {
-                if (wse.WebSocketErrorCode == WebSocketError.UnsupportedProtocol)
+                    // TODO: check tunnel access token expiration and try refresh it if it's expired.
+                    stream = await this.relayClient.CreateSessionStreamAsync(cts.Token);
+                    await this.relayClient.ConfigureSessionAsync(stream, isReconnect, cts.Token);
+
+                    stream = null;
+                    disconnectReason = SshDisconnectReason.None;
+                    isFinalAttempt = true;
```

Comment:
Is there a reason to set this local variable before returning?

Reply:
Yep, it's used in `finally` block that fires on `return` too.